### ISSUE TITLE
[memo] High memory consumption and the places of doubts

### ISSUE THREAD

**Issue by nabenabe0928** [automl/Auto-PyTorch#180]
I write down the current memory usage as a memo just in case when we encounter memory leak issues in the future.
This post is based on [the current implementation](https://github.com/automl/Auto-PyTorch/commit/a4e08e297dc298b53d5962f0f4b4e55b5e9ded8c).

When we run a dataset with the size of 300B, AutoPytorch consumes ~1.5GB and the followings are the major reason for the memory consumption:
|Source| Consumption [GB] | 
|-| :-:|
|Import modules| 0.35 |
|Dask Client | 0.35 |
|Logger | 0.4 |
| Running of context.Process in multiprocessing module | 0.4 |
| Model | 0 ~ inf |
|**Total**| 1.5 ~ inf |

When we run a dataset with the size of 300MB (400,000 instances x 80 features) such as Albert, AutoPytorch consumes ~1.7GB and the followings are the major reason for the memory consumption:
|Source| Consumption [GB] | 
|-| :-:|
|Import modules| 0.35|
|Dask Client | 0.35 |
|Logger | 0.4 |
| Dataset itself | 0.3 |
|self.categories in InputValidator| 0.3 |
| Running of context.Process in multiprocessing module | 0.4 |
| Model (e.g. LightGBM) | 0.4 ~ inf |
|**Total**| 2.5 ~ inf |

**Comment by ArlindKadra** [comment]
Interesting :), I think the analysis in the future should also be extended to the following datasets:
 
 https://archive.ics.uci.edu/ml/datasets/covertype  
 https://archive.ics.uci.edu/ml/datasets/HIGGS
 https://archive.ics.uci.edu/ml/datasets/Poker+Hand
 
 They proved tricky.

**Comment by nabenabe0928** [comment]
When we call the following, we consume 4GB even before the model instantiation:
```
python -m pytest test -vs -k <test_predict>
```

**Comment by nabenabe0928** [comment]
FYI, when we use [Optuna](https://github.com/optuna/optuna) with a tiny model, we consume around only 150MB.
This module is also thread safe.

```
import optuna


def objective(trial):
    x0 = trial.suggest_uniform('x0', -10, 10)
    x1 = trial.suggest_uniform('x1', -10, 10)
    return x0 ** 2 + x1 ** 2


if __name__ == '__main__':
    study = optuna.create_study()
    study.optimize(objective, n_trials=5000, n_jobs=4)
```

**Comment by nabenabe0928** [comment]
|Source| Consumption in covertype [GB] | Consumption in higgs [GB] | Consumption in pocker-hand [GB] | 
|-| :-:|:-:|:-:|
|Import modules| 0.35| | |
|Dask Client | 0.35 || |
|Logger (Thread safe) | 0.4 || |
| Dataset itself | 0.3 || |
|self.categories in InputValidator| 0.3 || |
| Running of context.Process in multiprocessing module | 0.4 || |
|**Total**| 2.5 ~ inf || |

**Comment by nabenabe0928** [comment]
This is from #259.

- [ ] We should not let the datamanager actively reside in memory when we are not using it. For example, there is no need to have a datamanager in [smbo](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/optimizer/smbo.py#L177).
- [ ] Also, after search has save the datamanager to disk, we can delete and garbage collect it.
- [ ] We also should datacollect and challenge the need of datamanager in the [evaluator](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/evaluation/tae.py#L171)
- [ ] We should improve the cross validation handling of the out of fold predictions. Rather than having a list that contains the OOF predicitons [here](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/evaluation/train_evaluator.py#L240) we should have a fixed array of n_samples created once at the beginning. OOF predictions from the k-fold model should be added smartly to this pre-existing array, something like `self.Y_optimization[test_indices] = opt_pred`. This ways predictions are sorted and can be used directly by ensemble selection without the need of saving this [array](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/evaluation/train_evaluator.py#L239)
- [ ] Calculating the train loss should be optional, not done by default [here](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/evaluation/train_evaluator.py#L257). We should prevent doing predict if not strictly needed.
- [ ] As reported already by @nabenabe0928 the biggest contribution comes from import files. In particular, just doing `import torch` consumes 2Gb of peak virtual memory and the majority of times this happens is for mypy typing. We should encapsulate these calls under `typing.TYPE_CHECKING` and only import the strictly needed class from pytorch.
